drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
 	return 0;
 }
 
-int usb_dc_ep_read_continue(uint8_t ep)
+/*
+ * @brief utility function, grouping control ep code, called by usb_dc_ep_read_continue
+ *
+ * @param [in] ep_state          ep0_out state
+ * @param [out] arm_out_endpoint set to true, if more data transactions are expected
+ *
+ * @return error code
+ */
+static int usb_dc_control_ep_read_continue(const struct udc_rpi_ep_state *const ep_state,
+	bool *const arm_out_endpoint)
 {
-	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
-	size_t bytes_received;
+	const struct usb_setup_packet *const setup = (const void *)&usb_dpram->setup_packet;
+
+	if (state.setup_available) {
+		LOG_DBG("EP0 setup (wLength=%d, is_to_device=%d)",
+			setup->wLength, usb_reqtype_is_to_device(setup));
+		if (setup->wLength != 0U) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint
+			 * exactly once, to either:
+			 * 1) in the to_device case, to receive the data (only if wLength is not 0)
+			 * 2) in the to_host case, to receive a 0-length status stage transfer
+			 *    (only valid if wLength is not 0)
+			 * Note that when wLength = 0, the status stage transfer is always an IN
+			 * type so we don't need to consider that case.
+			 */
+			*arm_out_endpoint = true;
+			state.control_out_ep_rcvd = 0;
+		}
+		state.setup_available = false;
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(USB_CONTROL_EP_OUT);
+
+		LOG_DBG("EP0_OUT received %d offset: %d", bytes_received, ep_state->read_offset);
+		if (usb_reqtype_is_to_device(setup)) {
+			if (state.control_out_ep_rcvd + ep_state->read_offset < setup->wLength) {
+				/* If no more data in the buffer, but we're still waiting
+				 * for more, start a new read transaction.
+				 */
+				if (bytes_received == ep_state->read_offset) {
+					state.control_out_ep_rcvd += ep_state->read_offset;
+					*arm_out_endpoint = true;
+				}
+			}
+		}
+	}
```

Comment:
It looks like new functionality or where does it come from? The stack can handle it transaction by transaction.

Reply:
- The outer if(), is indeed new functionality. The very first version of this commit only re-armed the endpoint once after the setup packet is received. It was okay for short to_device transactions, however when wLength is longer than the maximum packet size, the out endpoint will need to be re-armed multiple times, until the whole data stage is consumed. The necessity of this was not exposed with my initial testing using keyboard firmware and set locklight messages, however I noticed this happening while testing the RNDIS app.

  When I pushed the second version of this commit,  I added this functionality and I have updated the commit message to say "for EP0 it is only re-armed if more OUT transactions are expected for that SETUP transfer (be it data-stage or status-stage)". In my opinion this line covers why line 770 is there but if you would like me to add more detail to the commit message, or comments, let me know.

- The inner if() is existing functionality, it matches line 646 of the original file: https://github.com/zephyrproject-rtos/zephyr/pull/54030/files#diff-927fe840a66cec50c33e92952dbef313db723ed7a43bdea4603da8d3c312373dL646
About that if() I already had a short conversation with desowin, see here: https://github.com/zephyrproject-rtos/zephyr/pull/54030#discussion_r1090263428
I do not feel comfortable to remove functionality as part of this PR even if I believed it was unused.